src/poly/commitment/verifier.rs

```diff
@@ -62,174 +61,122 @@ impl<'a, C: CurveAffine> Guard<'a, C> {
-            *scalar *= &(*c);
-        }
-
-        for (scalar, base) in extra_scalars.iter().zip(extra_bases.iter()) {
-            msm.append_term(*scalar, *base);
-        }
-
-        // [c * v] U - [z1 * b] U
-        msm.append_term((*c * &v) + &(neg_z1 * &b), u);
-
-        // delta
-        msm.append_term(Field::one(), self.delta);
+/// Checks to see if an [`Proof`] is valid given the current `transcript`, and a
+/// point `x` that the polynomial commitment `P` opens purportedly to the value
+/// `v`. The provided `msm` should evaluate to the commitment `P` being opened.
+pub fn verify_proof<'a, C: CurveAffine, T: TranscriptRead<C>>(
+    params: &'a Params<C>,
+    mut msm: MSM<'a, C>,
+    transcript: &mut T,
+    x: C::Scalar,
+    v: C::Scalar,
+) -> Result<Guard<'a, C>, Error> {
+    let k = params.k as usize;
+
+    //     P - [v] G_0 + S * iota
+    //   + \sum(L_i * u_i^2) + \sum(R_i * u_i^-2)
+    msm.add_constant_term(-v);
+    let s_poly_commitment = transcript.read_point().map_err(|_| Error::OpeningError)?;
+
+    let iota = *ChallengeScalar::<C, ()>::get(transcript);
+    msm.append_term(iota, s_poly_commitment);
+
+    let z = *ChallengeScalar::<C, ()>::get(transcript);
+
+    let mut rounds = vec![];
+    for _ in 0..k {
+        // Read L and R from the proof and write them to the transcript
+        let l = transcript.read_point().map_err(|_| Error::OpeningError)?;
+        let r = transcript.read_point().map_err(|_| Error::OpeningError)?;
+
+        let challenge_packed = Challenge::get(transcript);
+        let challenge = *ChallengeScalar::<C, ()>::from(challenge_packed);
+
+        rounds.push((
+            l,
+            r,
+            challenge,
+            /* to be inverted */ challenge,
+            challenge_packed,
+        ));
+    }
 
-        // + [z1 - z2] H
-        msm.add_to_h_scalar(self.z1 - &self.z2);
+    rounds
+        .iter_mut()
+        .map(|&mut (_, _, _, ref mut challenge, _)| challenge)
+        .batch_invert();
 
-        let guard = Guard {
-            msm,
-            neg_z1,
-            allinv,
-            challenges_sq,
-            challenges_sq_packed,
-        };
+    let mut challenges = Vec::with_capacity(k);
+    let mut challenges_packed: Vec<Challenge> = Vec::with_capacity(k);
+    for (l, r, challenge, challenge_inv, challenge_packed) in rounds {
+        msm.append_term(challenge_inv, l);
+        msm.append_term(challenge, r);
 
-        Ok(guard)
+        challenges.push(challenge);
+        challenges_packed.push(challenge_packed);
     }
+
+    // Our goal is to open
+    //     msm - [v] G_0 + random_poly_commitment * iota
+    //   + \sum(L_i * u_i^2) + \sum(R_i * u_i^-2)
+    // at x to 0, by asking the prover to supply (a, h) such that it equals
+    //   = [a] (G + [b * z] U) + [h] H
+    // except that we wish for the prover to supply G as Commit(g(X); 1) so
+    // we must substitute to get
+    //   = [a] ((G - H) + [b * z] U) + [h] H
```

Comment:
I don't understand how this is a substitution into the previous line.

Reply:
It's the left hand side of the equality, `msm - [v] G_0 + random_poly_commitment * iota + \sum(L_i * u_i^2) + \sum(R_i * u_i^-2)`.

Reply:
The prover will not be giving us `G` but rather `G + H` (the blinding factor is 1), so we have to subtract `H` from what they give us. We do this because auxillary commitments always have a blinding factor of 1 in the PLONK API, ensuring that under no circumstance will points at infinity emerge inside of a recursive circuit.